tools-v2/pkg/cli/command/curvebs/delete/peer/peer.go

```diff
@@ -0,0 +1,202 @@
+ */
+
+package peer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc"
+
+	cmderror "github.com/opencurve/curve/tools-v2/internal/error"
+	cobrautil "github.com/opencurve/curve/tools-v2/internal/utils"
+	basecmd "github.com/opencurve/curve/tools-v2/pkg/cli/command"
+	"github.com/opencurve/curve/tools-v2/pkg/config"
+	"github.com/opencurve/curve/tools-v2/pkg/output"
+	"github.com/opencurve/curve/tools-v2/proto/proto/cli"
+	"github.com/opencurve/curve/tools-v2/proto/proto/common"
+)
+
+const (
+	deleteExample = `$ curve bs delete peer 127.0.0.1:8200:0 --logicalpoolid=1 --copysetid=10001
+ --peers=127.0.0.1:8200:0,127.0.0.1:8201:1,127.0.0.1:8202:2 --rpcretrytimes=1 --rpctimeout=10s`
+)
+
+// RPCClient the rpc client for the rpc function RemovePeer
+type RPCClient struct {
+	Info    *basecmd.Rpc
+	Request *cli.RemovePeerRequest
+	cli     cli.CliServiceClient
+}
+
+func (rpp *RPCClient) NewRpcClient(cc grpc.ClientConnInterface) {
+	rpp.cli = cli.NewCliServiceClient(cc)
+}
+
+func (rpp *RPCClient) Stub_Func(ctx context.Context) (interface{}, error) {
+	return rpp.cli.RemovePeer(ctx, rpp.Request)
+}
+
+var _ basecmd.RpcFunc = (*RPCClient)(nil) // check interface
+
+// Command the command to perform remove peer.
+type Command struct {
+	basecmd.FinalCurveCmd
+
+	// request parameters
+	opts          Options
+	logicalPoolID uint32
+	copysetID     uint32
+
+	conf       Configuration
+	removePeer *common.Peer
+}
+
+var _ basecmd.FinalCurveCmdFunc = (*Command)(nil) // check interface
+// NewCommand ...
+func NewCommand() *cobra.Command {
+	cCmd := &Command{
+		FinalCurveCmd: basecmd.FinalCurveCmd{
+			Use:     "peer",
+			Short:   "delete the peer from the copyset",
+			Example: deleteExample,
+		},
+	}
+	basecmd.NewFinalCurveCli(&cCmd.FinalCurveCmd, cCmd)
+	return cCmd.Cmd
+}
+
+func (cCmd *Command) AddFlags() {
+	config.AddRpcRetryTimesFlag(cCmd.Cmd)
+	config.AddRpcTimeoutFlag(cCmd.Cmd)
+
+	config.AddBSLogicalPoolIdFlag(cCmd.Cmd)
+	config.AddBSCopysetIdFlag(cCmd.Cmd)
+
+	config.AddBSPeersConfFlag(cCmd.Cmd)
+}
+
+func (cCmd *Command) Init(cmd *cobra.Command, args []string) error {
+	cCmd.opts = Options{}
+
+	cCmd.SetHeader([]string{cobrautil.ROW_LEADER, cobrautil.ROW_PEER, cobrautil.ROW_COPYSET, cobrautil.ROW_RESULT, cobrautil.ROW_REASON})
+	cCmd.TableNew.SetAutoMergeCellsByColumnIndex(cobrautil.GetIndexSlice(
+		cCmd.Header, []string{},
+	))
+
+	var err error
+	cCmd.opts.Timeout = config.GetFlagDuration(cCmd.Cmd, config.RPCTIMEOUT)
+	cCmd.opts.RetryTimes = config.GetFlagInt32(cCmd.Cmd, config.RPCRETRYTIMES)
+
+	cCmd.copysetID, err = config.GetBsFlagUint32(cCmd.Cmd, config.CURVEBS_COPYSET_ID)
+	if err != nil {
+		return err
+	}
+	cCmd.logicalPoolID, err = config.GetBsFlagUint32(cCmd.Cmd, config.CURVEBS_LOGIC_POOL_ID)
+	if err != nil {
+		return err
+	}
+
+	// parse peers config
+	curConf := config.GetBsFlagString(cCmd.Cmd, config.CURVEBS_PEERS_ADDRESS)
+	c, err := ParseConfiguration(curConf)
```

Comment:
Why not use stringslice? In this way, there is no need to do parsing of peers.

Reply:
the peers format is `address:id,address:id` which are different with normal address slice. The parse process is alse necessary.

Reply:
> the peers format is `address:id,address:id` which are different with normal address slice. The parse process is alse necessary.

stringslice can handle this situation, and the relevant processing flow can be moved to obtain the flag.

Reply:
I changed to `GetBsFlagStringSlice`, the function `ParseConfiguration` is remained to transfer the parameter `ip:port:index` to struct https://github.com/opencurve/curve/blob/c804338339eda0adc78451fc4a750e0bf1135054/proto/common.proto#L27